File: javascript/roman-numerals/roman-numerals.js
const ROMAN_NUMERALS = {
  1: 'I',
  4: 'IV',
  5: 'V',
  9: 'IX',
  10: 'X',
  40: 'XL',
  50: 'L',
  90: 'XC',
  100: 'C',
  400: 'CD',
  500: 'D',
  900: 'CM',
  1000: 'M'
};

function toRoman(n, romanNumeral = '') {
  var leftOver = 0;

  while (n > 0) {
    if (ROMAN_NUMERALS[n]) {
      romanNumeral += ROMAN_NUMERALS[n];
      n -= n;
    } else {
      leftOver += 1;
      n -= 1;
    }
  }

  if (leftOver > 0) { return toRoman(leftOver, romanNumeral); }

  return romanNumeral;
};

module.exports = toRoman;

File: ruby/roman-numerals/roman_numerals.rb
class Integer
  ROMAN_NUMERALS = {
    1000 => 'M',
    900  => 'CM',
    500  => 'D',
    400  => 'CD',
    100  => 'C',
    90   => 'XC',
    50   => 'L',
    40   => 'XL',
    10   => 'X',
    9    => 'IX',
    5    => 'V',
    4    => 'IV',
    1    => 'I'
  }

  def to_roman
    n = self
    roman_numeral = ''

    ROMAN_NUMERALS.each do |arabic, roman|
      while n >= arabic
        n -= arabic
        roman_numeral += roman
      end
    end

    roman_numeral
  end
end

class BookKeeping
  VERSION = 2
end
